### ISSUE TITLE
fatal error: concurrent map writes

### ISSUE THREAD

**Issue by james-laing** [PureStorage-OpenConnect/pure-fa-openmetrics-exporter#85]
Exporter exits with `fatal error: concurrent map writes` error.

While a workaround has been suggested:

**Binary**
Running as a service in Linux we can include `Restart=always` and `RestartSec=1`.

```
[Service]
Type=simple
Restart=always
RestartSec=1
ExecStart=/usr/bin/PureStorage-OpenConnect/pure-fa-openmetrics-exporter/out/bin/pure-fa-om-exporter --port 9490
```

**Docker** 
https://github.com/PureStorage-OpenConnect/pure-fa-openmetrics-exporter/blob/master/extra/grafana/README.md#check-docker-container-is-running
`docker run --restart unless-stopped`

We should really focus on finding the root cause of this issue to improve stability.

**Comment by james-laing** [comment]
Can we reopen this issue as it is still an outstanding and common issue.

**Comment by chrroberts-pure** [comment]
reopening - while looking there seems to be a number of malformed if statements on error code 401 in the rest-client.

i have a branch open - will draft a PR. I am currently doing testing in a lab for longterm use where this error was occurring.

**Comment by chrroberts-pure** [comment]
This is supported by the stacktrace as well with the error `purestorage/fa-openmetrics-exporter/internal/rest-client.(*FAClient).RefreshSession(0x14000397440)`

````
2023/11/01 14:17:20 Start Pure FlashArray exporter v1.0.9 on 0.0.0.0:9490
fatal error: concurrent map writes

goroutine 317 [running]:
net/textproto.MIMEHeader.Set(...)
	/usr/local/go/src/net/textproto/header.go:22
net/http.Header.Set(...)
	/usr/local/go/src/net/http/header.go:40
github.com/go-resty/resty/v2.(*Client).SetHeader(...)
	/Users/jlaing/go/pkg/mod/github.com/go-resty/resty/v2@v2.7.0/client.go:195
purestorage/fa-openmetrics-exporter/internal/rest-client.(*FAClient).RefreshSession(0x14000397440)
	/Users/jlaing/git/pure/pure-fa-openmetrics-exporter/internal/rest-client/flasharray_client.go:119 +0x3f0
purestorage/fa-openmetrics-exporter/internal/rest-client.(*FAClient).GetHardware(0x14000397440)
	/Users/jlaing/git/pure/pure-fa-openmetrics-exporter/internal/rest-client/hardware.go:36 +0x2d0
purestorage/fa-openmetrics-exporter/internal/openmetrics-exporter.(*HardwareCollector).Collect(0x1400006a8c0, 0x140004b2f58?)
	/Users/jlaing/git/pure/pure-fa-openmetrics-exporter/internal/openmetrics-exporter/hardware_collector.go:21 +0x34
github.com/prometheus/client_golang/prometheus.(*Registry).Gather.func1()
	/Users/jlaing/go/pkg/mod/github.com/prometheus/client_golang@v1.16.0/prometheus/registry.go:457 +0xc0
created by github.com/prometheus/client_golang/prometheus.(*Registry).Gather
	/Users/jlaing/go/pkg/mod/github.com/prometheus/client_golang@v1.16.0/prometheus/registry.go:547 +0x9f4
````

**Comment by chrroberts-pure** [comment]
closed with #127
